test/HTML5/Parser/DOMTreeBuilderTest.php

```diff
@@ -58,6 +58,80 @@ public function testDocument()
         $this->assertEquals('http://www.w3.org/1999/xhtml', $doc->documentElement->namespaceURI);
     }
 
+    public function testBareAmpersand()
+    {
+        $html = "<!doctype html>
+        <html>
+            <body> 
+                <img src='a&b' />
+                <img src='a&=' />
+                <img src='a&=c' />
+                <img src='a&=9' />
+            </body>
+        </html>";
+        $doc = $this->parse($html);
+
+        $this->assertEmpty($this->errors);
+        $this->assertXmlStringEqualsXmlString('
+        <!DOCTYPE html>
+        <html xmlns="http://www.w3.org/1999/xhtml"><body> 
+                <img src="a&amp;b"/>
+                <img src="a&amp;="/>
+                <img src="a&amp;=c"/>
+                <img src="a&amp;=9"/>
```

Comment:
Forgive me if I'm missing something here but, shouldn't the written attribute be `&` instead of `&amp;` since that was the input and it's within an attribute? I've not looked at the spec (or worked on this) in awhile so I could be off here.

Reply:
the specs says:

> **If the character reference was consumed as part of an attribute** (return state is either attribute value (double-quoted) state, attribute value (single-quoted) state or attribute value (unquoted) state), and the last character matched is not a U+003B SEMICOLON character (;), and the next input character is either a U+003D EQUALS SIGN character (=) or an alphanumeric ASCII character, then, **for historical reasons**, switch to the character reference end state.

My interpretation if it is that the correct form is `&amp;` but for "historical reasons" a mere  `&` is accepted. 
The current implementation is able to parse correctly `&`, but when serializing it back, it fixes it transforming it into `&amp;`.

Does it make sense?

Reply:
To me it does. Other "accepted syntax variations" are normalized as well when output is written.

Reply:
This makes sense. I was curious about browser handling and compatibility. Quite often developers use an `&`. For reference, it turns out that they translate `&amp;` to `&` when calling the link. There used to be issues with this in older browser (e.g., `amp;` being prepended to the next key/value pair in the query. I think this is fixed in all modern browser.